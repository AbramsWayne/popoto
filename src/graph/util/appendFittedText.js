import * as d3 from "d3";

var CONTEXT_2D = document.createElement("canvas").getContext("2d");
var DEFAULT_CANVAS_LINE_HEIGHT = 12;

export function measureTextWidth(text) {
    return CONTEXT_2D.measureText(text).width;
}

/**
 * Inspired by https://beta.observablehq.com/@mbostock/fit-text-to-circle
 * TODO: Clean getLines return and corresponding data.
 * @param text
 * @returns {*}
 */
export function getLines(text) {
    if (text === undefined || text === null) {
        return {lines: []}
    }

    var text = String(text);

    var words = text.split(/\s+/g); // To hyphenate: /\s+|(?<=-)/
    if (!words[words.length - 1]) words.pop();
    if (!words[0]) words.shift();

    var targetWidth = Math.sqrt(measureTextWidth(text.trim()) * DEFAULT_CANVAS_LINE_HEIGHT);

    var line;
    var lineWidth0 = Infinity;
    var lines = [];

    for (var i = 0, n = words.length; i < n; ++i) {
        var lineText1 = (line ? line.text + " " : "") + words[i];
        var lineWidth1 = measureTextWidth(lineText1);
        if ((lineWidth0 + lineWidth1) / 2 < targetWidth) {
            line.width = lineWidth0 = lineWidth1;
            line.text = lineText1;
        } else {
            lineWidth0 = measureTextWidth(words[i]);
            line = {width: lineWidth0, text: words[i]};
            lines.push(line);
        }
    }

    var textRadius = 0;

    for (var i = 0, n = lines.length; i < n; ++i) {
        var dy = (Math.abs(i - n / 2 + 0.5) + 0.5) * DEFAULT_CANVAS_LINE_HEIGHT;
        var dx = lines[i].width / 2;
        textRadius = Math.max(textRadius, Math.sqrt(dx * dx + dy * dy));
    }

    return {
        "lines": lines.map(function (d) {
            return {"text": d.text, "linesLength": lines.length}
        }),
        "textRadius": textRadius
    }
}

/**
 * Create the text representation of a node by slicing the text into lines to fit the node.
 *
 * @param selection
 * @param textParam
 * @param radiusParam
 * @param classParam
 */
export default function appendFittedText(selection, textParam, radiusParam, classParam) {
    var getRadius = typeof radiusParam === "function"
        ? radiusParam
        : function () {
            return radiusParam
        };

    var getText = typeof textParam === "function"
        ? textParam
        : function () {
            return textParam
        };

    var getClass = typeof classParam === "function"
        ? classParam
        : function () {
            return classParam?classParam:""
        };

    selection.each(function (d) {
        d3.select(this)
            .selectAll(".fitted-text")
            .data([{}])
            .enter()
            .append("text")
            .attr("class", getClass(d) + " fitted-text")
            .attr("style", "text-anchor: middle; font: 10px sans-serif");
    });

    var text = selection.select(".fitted-text");

    text.datum(function (d) {
        var lines = getLines(getText(d));
        d.lines = lines.lines;
        d.textRadius = lines.textRadius;
        return d
    });

    text.selectAll("tspan")
        .data(function (d) {return d.lines})
        .enter()
        .append("tspan")
        .attr("x", 0)
        .attr("y", function (d,i) { return (i - d.linesLength / 2 + 0.8) * DEFAULT_CANVAS_LINE_HEIGHT})
        .text(function(d) {return d.text});

    text.attr("transform", function (d) {
        var scale = 1;
        if (d.textRadius !== 0 && d.textRadius) {
            scale = getRadius(d) / d.textRadius;
        }
        return "translate(" + 0 + "," + 0 + ")" + " scale(" + scale + ")"
    });
}