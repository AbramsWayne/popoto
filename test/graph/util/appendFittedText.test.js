import * as d3 from "d3"
import appendFittedText from '../../../src/graph/util/appendFittedText'
import {measureTextWidth, getLines} from '../../../src/graph/util/appendFittedText'

describe("measureTextWidth", function () {
    test("Should measure undefined text correctly", () => {
        expect(measureTextWidth()).toMatchSnapshot();
    });

    test("Should measure empty text correctly", () => {
        expect(measureTextWidth("")).toMatchSnapshot();
    });

    test("Should measure text correctly", () => {
        expect(measureTextWidth("mockedTextValue")).toMatchSnapshot();
    });
});

describe("getLines", function () {

    test("Should works with undefined as text", () => {
        const lines = getLines();
        expect(lines).toMatchSnapshot();
    });

    test("Should works with empty string as text", () => {
        const lines = getLines("");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with simple word as text", () => {
        const lines = getLines("mockedTextValue");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple words on same line", () => {
        const lines = getLines("mocked text a value");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple words as text", () => {
        const lines = getLines("mocked text value");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple spaces and multiple words as text", () => {
        const lines = getLines("        mocked   text       value           ");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with integer as text", () => {
        const lines = getLines(0);
        expect(lines).toMatchSnapshot();
    });

});

describe("appendFittedText", function () {

    describe('Should works with text, radius, class parameters as value without data in selection', () => {
        beforeEach(() => {
            const textMockValue = "mocked_text_value";
            const radiusMockValue = 50;
            const cssClassMockValue = "mocked_css_class_value";

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '</div>';
            appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should works with text, radius, class parameters as function without data in selection', () => {
        beforeEach(() => {
            const textMockValue = function (d) {
                return "mocked_text_value"
            };
            const radiusMockValue = function (d) {
                return 50
            };
            const cssClassMockValue = function (d) {
                return "mocked_css_class_value"
            };

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '</div>';

            appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should works with text, radius, class parameters as function with data in selection', () => {
        beforeEach(() => {
            const textMockValue = function (d) {
                return d.textMockValue
            };
            const radiusMockValue = function (d) {
                return d.radiusMockValue
            };
            const cssClassMockValue = function (d) {
                return d.cssClassMockValue
            };

            const data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
            ];

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '<g>' +
                '</g>' +
                '</div>';

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });
});