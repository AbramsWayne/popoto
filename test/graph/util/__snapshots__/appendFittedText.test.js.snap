// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appendFittedText Should works with text, radius, class parameters as function with data in selection Should render correctly 1`] = `"<div><g><text class=\\"mocked_css_class_value fitted-text\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"mocked_css_class_value2 fitted-text\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48074017006186526)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value2</tspan></text></g></div>"`;

exports[`appendFittedText Should works with text, radius, class parameters as function without data in selection Should render correctly 1`] = `"<div><g><text class=\\"mocked_css_class_value fitted-text\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`appendFittedText Should works with text, radius, class parameters as value without data in selection Should render correctly 1`] = `"<div><g><text class=\\"mocked_css_class_value fitted-text\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`extractLines Should works with empty string as text 1`] = `Array []`;

exports[`extractLines Should works with integer as text 1`] = `
Array [
  Object {
    "text": "0",
    "width": 12,
  },
]
`;

exports[`extractLines Should works with multiple spaces and multiple words as text 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "value",
    "width": 60,
  },
]
`;

exports[`extractLines Should works with multiple words as text 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "value",
    "width": 60,
  },
]
`;

exports[`extractLines Should works with multiple words on same line 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "a value",
    "width": 84,
  },
]
`;

exports[`extractLines Should works with simple word as text 1`] = `
Array [
  Object {
    "text": "mockedTextValue",
    "width": 180,
  },
]
`;

exports[`extractLines Should works with undefined as text 1`] = `Array []`;

exports[`measureTextWidth Should measure empty text correctly 1`] = `0`;

exports[`measureTextWidth Should measure text correctly 1`] = `180`;

exports[`measureTextWidth Should measure undefined text correctly 1`] = `0`;
